File: web/src/api/search_engine.ts
import request, { type ListQuery } from './request'

export type SearchEngineProvider = 'tavily' | 'serpapi' | 'aliyun-iqs'

export interface SearchEngineConfig {
  id: number
  provider: SearchEngineProvider
  name: string
  base_url: string
  api_key_set: boolean
  enabled: boolean
  created_at?: string
  updated_at?: string
}

export interface UpdateSearchEngineConfigReq {
  provider?: SearchEngineProvider
  name?: string
  base_url?: string
  api_key?: string
  enabled?: boolean
}

export interface CreateSearchEngineConfigReq {
  provider: SearchEngineProvider
  name: string
  base_url: string
  api_key?: string
  enabled: boolean
}

export interface SearchEngineResult {
  title: string
  url: string
  snippet: string
}

export interface SearchEngineTestResp {
  query: string
  provider: string
  results: SearchEngineResult[]
}

export interface TestSearchEngineConfigReq {
  id?: number
  query: string
  limit?: number
  provider: SearchEngineProvider
  name: string
  base_url: string
  api_key?: string
}

export const searchEngineApi = {
  list: (params?: ListQuery) => request.get('/search-engines', { params }),
  get: (id: number) => request.get(`/search-engines/${id}`),
  create: (data: CreateSearchEngineConfigReq) => request.post('/search-engines', data),
  update: (id: number, data: UpdateSearchEngineConfigReq) => request.put(`/search-engines/${id}`, data),
  delete: (id: number) => request.delete(`/search-engines/${id}`),
  test: (id: number, data: { query: string; limit?: number }) => request.post(`/search-engines/${id}/test`, data),
  testConfig: (data: TestSearchEngineConfigReq) => request.post('/search-engines/test', data),
}

File: internal/model/search_engine.go
package model

import "time"

type SearchEngineProvider string

const (
	SearchEngineTavily    SearchEngineProvider = "tavily"
	SearchEngineSerpAPI   SearchEngineProvider = "serpapi"
	SearchEngineAliyunIQS SearchEngineProvider = "aliyun-iqs"
)

type SearchEngineConfig struct {
	ID        int64                `json:"id" gorm:"primaryKey;autoIncrement"`
	Provider  SearchEngineProvider `json:"provider" gorm:"size:50;not null"`
	Name      string               `json:"name" gorm:"size:100;not null"`
	BaseURL   string               `json:"base_url" gorm:"size:500;not null"`
	APIKey    string               `json:"api_key,omitempty" gorm:"size:500"`
	Enabled   bool                 `json:"enabled" gorm:"not null;default:false"`
	CreatedAt time.Time            `json:"created_at"`
	UpdatedAt time.Time            `json:"updated_at"`
}

type SearchEngineConfigResp struct {
	ID        int64                `json:"id"`
	Provider  SearchEngineProvider `json:"provider"`
	Name      string               `json:"name"`
	BaseURL   string               `json:"base_url"`
	APIKeySet bool                 `json:"api_key_set"`
	Enabled   bool                 `json:"enabled"`
	CreatedAt time.Time            `json:"created_at,omitzero"`
	UpdatedAt time.Time            `json:"updated_at,omitzero"`
}

type UpdateSearchEngineConfigReq struct {
	Provider *SearchEngineProvider `json:"provider,omitzero"`
	Name     *string               `json:"name,omitzero"`
	BaseURL  *string               `json:"base_url,omitzero"`
	APIKey   *string               `json:"api_key,omitzero"`
	Enabled  *bool                 `json:"enabled,omitzero"`
}

type CreateSearchEngineConfigReq struct {
	Provider SearchEngineProvider `json:"provider"`
	Name     string               `json:"name"`
	BaseURL  string               `json:"base_url"`
	APIKey   string               `json:"api_key"`
	Enabled  bool                 `json:"enabled"`
}

type TestSearchEngineReq struct {
	Query string `json:"query"`
	Limit int    `json:"limit,omitzero"`
}

type TestSearchEngineConfigReq struct {
	ID       int64                `json:"id,omitzero"`
	Query    string               `json:"query"`
	Limit    int                  `json:"limit,omitzero"`
	Provider SearchEngineProvider `json:"provider,omitzero"`
	Name     string               `json:"name,omitzero"`
	BaseURL  string               `json:"base_url,omitzero"`
	APIKey   string               `json:"api_key,omitzero"`
}
